server/src/main/java/org/elasticsearch/action/admin/indices/create/CreateIndexRequest.java

```diff
@@ -389,15 +402,20 @@ public CreateIndexRequest source(Map<String, ?> source, DeprecationHandler depre
                 aliases((Map<String, Object>) entry.getValue());
             } else {
                 // maybe custom?
-                IndexMetaData.Custom proto = IndexMetaData.lookupPrototype(name);
-                if (proto != null) {
+                if (entry.getValue() instanceof Map) {
+                    if (customsRegistry == null) {
+                        // Custom registry wasn't specified - we are in client mode let's try loading it from SPI
+                        customsRegistry = new NamedXContentRegistry(getProvidedNamedXContents());
```

Comment:
I'm not sure to understand what you want to achieve here (though I understand the proto based serialization removal).

I don't think we want a request object to create a `NamedXContentRegistry` with only some provided names xcontents as it won't contain all named xcontent parsers but only the ones that provide and implement `NamedXContentProvider`. 

The NamedXContentRegistry is usually instantiated once on the server or client side and it must contain all required named xcontent parsers. This list is potentially enriched with named xcontent parsers provided by modules or plugins that implement `NamedXContentProvider`. But in any case, the NamedXContentRegistry is instantiated once on server or client side with everything it needs.

On server side the registry is instantiated in the [Node class](https://github.com/elastic/elasticsearch/blob/db6e8c736d92582fe56024993450ce0a987b498d/server/src/main/java/org/elasticsearch/node/Node.java#L394-L401) and in the [RestHighLevelClient class](https://github.com/elastic/elasticsearch/blob/db6e8c736d92582fe56024993450ce0a987b498d/client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java#L236-L238) for an example of client side instanciation.

In your case, I guess you'll have to plumb the `NamedXContentRegistry` down to where you want to parse this content. I think it's also possible to retrieve a ref on the current registry from a XContentParser instance.

Reply:
Thanks! As per our discussion, I have removed the automatic `NamedXContentRegistry` discovery and replaced with passing it to `source(...)` method as parameters.